client/vaultclient/vaultclient.go

```diff
@@ -0,0 +1,933 @@
+	// Client's region
+	region string
+
+	// The node in which this vault client is running in
+	node *structs.Node
+
+	// running indicates if the renewal loop is active or not
+	running bool
+
+	// connEstablished marks whether the connection to vault was successful
+	// or not
+	connEstablished bool
+
+	// tokenData is the data of the passed VaultClient token
+	token *tokenData
+
+	// API client to interact with vault
+	client *vaultapi.Client
+
+	// Channel to notify heap modifications to the renewal loop
+	updateCh chan struct{}
+
+	// Channel to trigger termination of renewal loop
+	stopCh chan struct{}
+
+	// Min-Heap to keep track of both tokens and leases
+	heap *vaultClientHeap
+
+	// Configuration to connect to vault
+	config *config.VaultConfig
+
+	lock   sync.RWMutex
+	logger *log.Logger
+
+	rpcHandler clientconfig.RPCHandler
+	rpcProxy   *rpcproxy.RPCProxy
+	connPool   *nomad.ConnPool
+}
+
+// tokenData holds the relevant information about the Vault token passed to the
+// client.
+type tokenData struct {
+	CreationTTL int      `mapstructure:"creation_ttl"`
+	TTL         int      `mapstructure:"ttl"`
+	Renewable   bool     `mapstructure:"renewable"`
+	Policies    []string `mapstructure:"policies"`
+	Role        string   `mapstructure:"role"`
+	Root        bool
+}
+
+// Request object for renewals. This can be used for both token renewals and
+// secret's lease renewals.
+type vaultClientRenewalRequest struct {
+	// Channel into which any renewal error will be sent down to
+	errCh chan error
+
+	// This can either be a token or a lease identifier
+	id string
+
+	// Duration for which the token or lease should be renewed for
+	increment int
+
+	// Indicates whether the 'id' field is a token or not
+	isToken bool
+}
+
+// Element representing an entry in the renewal heap
+type vaultClientHeapEntry struct {
+	req   *vaultClientRenewalRequest
+	next  time.Time
+	index int
+}
+
+// Wrapper around the actual heap to provide additional symantics on top of
+// functions provided by the heap interface. In order to achieve that, an
+// additional map is placed beside the actual heap. This map can be used to
+// check if an entry is already present in the heap.
+type vaultClientHeap struct {
+	heapMap map[string]*vaultClientHeapEntry
+	heap    vaultDataHeapImp
+}
+
+// Data type of the heap
+type vaultDataHeapImp []*vaultClientHeapEntry
+
+// NewVaultClient returns a new vault client from the given config.
+func NewVaultClient(node *structs.Node, region string, config *config.VaultConfig,
+	logger *log.Logger, rpcHandler clientconfig.RPCHandler, connPool *nomad.ConnPool,
+	rpcProxy *rpcproxy.RPCProxy) (*vaultClient, error) {
+	if !config.Enabled {
+		return nil, nil
+	}
+
+	if node == nil {
+		return nil, fmt.Errorf("nil node")
+	}
+
+	if region == "" {
+		return nil, fmt.Errorf("missing region")
+	}
+
+	if config == nil {
+		return nil, fmt.Errorf("nil vault config")
+	}
+
+	// Creation of a vault client requires a token
```

Comment:
The client shouldn't need to be given a Vault token. Unwrapping doesn't require a token and then when it is renewing it can set the token to be that of the unwrapped token